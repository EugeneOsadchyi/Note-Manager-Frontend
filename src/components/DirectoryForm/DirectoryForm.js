import React, { Component } from 'react';
import PropTypes from 'prop-types';

import updateObject from '../../utils/updateObject';
import { checkValidity } from '../../utils/validation';

import Input from '../UI/Input';
import Button from '../UI/Button';

class DirectoryForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      directoryForm: {
        name: {
          elementType: 'input',
          elementConfig: {
            type: 'text',
            placeholder: 'Directory Name',
          },
          value: props.directoryName,
          validation: {
            required: true,
          },
          valid: false,
        },
      },
      formIsValid: false,
    };
  }

  saveClickedHandler(event) {
    event.preventDefault();

    const {
      directoryForm,
    } = this.state;

    const {
      directoryId,
      saveClicked,
    } = this.props;

    const formData = {};

    Object.keys(directoryForm).forEach((formElementIdentifier) => {
      formData[formElementIdentifier] = directoryForm[formElementIdentifier].value;
    });

    const directory = {
      directoryId,
      directoryData: formData,
    };

    saveClicked(directory);
  }

  inputChangedHandler(event, inputIdentifier) {
    const { directoryForm } = this.state;

    const updatedFormElement = updateObject(directoryForm[inputIdentifier], {
      value: event.target.value,
      valid: checkValidity(event.target.value, directoryForm[inputIdentifier].validation),
      touched: true,
    });

    const updatedDirectoryForm = updateObject(directoryForm, {
      [inputIdentifier]: updatedFormElement,
    });

    let formIsValid = true;
    Object.keys(updatedDirectoryForm).forEach((fieldName) => {
      formIsValid = updatedDirectoryForm[fieldName].valid && formIsValid;
    });

    this.setState({ directoryForm: updatedDirectoryForm, formIsValid });
  }

  render() {
    const {
      cancelClicked,
    } = this.props;

    const {
      directoryForm,
      formIsValid,
    } = this.state;

    const formElementsArray = [];
    Object.keys(directoryForm).forEach((key) => {
      formElementsArray.push({
        id: key,
        config: directoryForm[key],
      });
    });

    return (
      <form onSubmit={this.saveClickedHandler}>
        {formElementsArray.map(formElement => (
          <Input
            key={formElement.id}
            elementType={formElement.config.elementType}
            elementConfig={formElement.config.elementConfig}
            value={formElement.config.value}
            invalid={!formElement.config.valid}
            shouldValidate={formElement.config.validation}
            touched={formElement.config.touched}
            changed={event => this.inputChangedHandler(event, formElement.id)}
          />
        ))}
        <Button btnType="btn-cancel" clicked={cancelClicked}>
          Cancel
        </Button>
        <Button btnType="Success btn-save" disabled={!formIsValid}>
          Save
        </Button>
      </form>
    );
  }
}

DirectoryForm.propTypes = {
  directoryId: PropTypes.number,
  directoryName: PropTypes.string,
  saveClicked: PropTypes.func.isRequired,
  cancelClicked: PropTypes.func.isRequired,
};

DirectoryForm.defaultProps = {
  directoryId: undefined,
  directoryName: '',
};

export default DirectoryForm;
