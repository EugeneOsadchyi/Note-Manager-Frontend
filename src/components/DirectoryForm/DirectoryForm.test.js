import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import DirectoryForm from './DirectoryForm';

describe('DirectoryForm', () => {
  const mockSaveClicked = jest.fn();
  const mockCancelClicked = jest.fn();

  const directoryName = 'New Directory';

  const props = {
    directoryId: undefined,
    directoryName: '',
    saveClicked: mockSaveClicked,
    cancelClicked: mockCancelClicked,
  };

  let directoryForm = shallow(<DirectoryForm {...props} />);

  beforeEach(() => {
    mockSaveClicked.mockReset();
    mockCancelClicked.mockReset();
  });

  it('renders properly', () => {
    expect(shallowToJson(directoryForm)).toMatchSnapshot();
  });

  it('contains input component', () => {
    expect(directoryForm.find('Input').exists()).toBe(true);
  });

  it('contains button components', () => {
    expect(directoryForm.find('Button').length).toEqual(2);
  });

  describe('when directory name is set', () => {
    beforeEach(() => {
      props.directoryName = directoryName;
      directoryForm = mount(<DirectoryForm {...props} />);
    });

    it('input component pre-files with directory name', () => {
      expect(directoryForm.find('Input').props().value).toEqual(directoryName);
    });

    describe('and save button was clicked', () => {
      beforeEach(() => {
        directoryForm.find('.btn-save').simulate('click');
      });

      it('dispatches `saveClicked()` it received from the props', () => {
        expect(mockSaveClicked).toHaveBeenCalled();
      });
    });
  });

  describe('when directory name is not set', () => {
    beforeEach(() => {
      props.directoryName = '';
      directoryForm = mount(<DirectoryForm {...props} />);
    });

    describe('and save button was clicked', () => {
      beforeEach(() => {
        directoryForm.find('.btn-save').simulate('click');
      });

      it('does not dispatch `saveClicked()` it received from the props', () => {
        expect(mockSaveClicked).not.toHaveBeenCalled();
      });
    });
  });
});
