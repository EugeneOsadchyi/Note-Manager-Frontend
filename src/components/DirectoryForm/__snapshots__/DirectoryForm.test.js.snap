// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectoryForm renders properly 1`] = `
<form
  onSubmit={[Function]}
>
  <Input
    changed={[Function]}
    elementConfig={
      Object {
        "placeholder": "Directory Name",
        "type": "text",
      }
    }
    elementType="input"
    invalid={true}
    key="name"
    shouldValidate={
      Object {
        "required": true,
      }
    }
    touched={false}
    value=""
  />
  <Button
    btnType="btn-cancel"
    clicked={[Function]}
    disabled={false}
  >
    Cancel
  </Button>
  <Button
    btnType="Success btn-save"
    clicked={[Function]}
    disabled={true}
  >
    Save
  </Button>
</form>
`;
